text/0000-destructor-guarantees.md

```diff
@@ -0,0 +1,176 @@
+> addition to `thread::scoped`.
+
+[scoped-bug]: https://github.com/rust-lang/rust/issues/24292
+[forget-bug]: https://github.com/rust-lang/rust/issues/24456
+[dtor-comment]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93505374
+[dtor-bug1]: https://github.com/rust-lang/rust/issues/14875
+[dtor-bug2]: https://github.com/rust-lang/rust/issues/16135
+[drain-bug]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93513451
+
+Previously, Rust provided no guarantee that any destructors for any types will
+run. However, some of the current APIs (namely `thread::scoped`) were designed
+without keeping this in mind, and can be made memory unsafe through leaking.
+
+This RFC proposes a small, orthogonal feature to allow wondrous RAII-based APIs
+like `thread::scoped` to exist in safe Rust, with minimal breaking changes in
+the pre-1.0 release.
+
+## Narrowing Goals and Dispelling Fear, Uncertainty, and Doubt
+
+As this issue touched a fairly popular and touted API (`thread::scoped`) there
+has naturally been a gigantic amount of discussion and ideation surrounding it.
+
+Some of the proposed solutions to this problem have been far too restrictive,
+such as banning `Rc` cycles entirely, making `Rc` unsafe, or having `Rc`
+require `'static`. This had lead to a certain amount of FUD about the validity
+of all non-safe-`mem::forget` solutions. This RFC is here to dispel this illusion.
+
+### This Proposal Will Not:
+
+ - Prevent *all* destructors from not running.
+ - Prevent destructors from not running in the face of process aborts.
+ - Prevent destructors from not running for `'static` data.
+ - Prevent `Rc` or `Arc` cycles, or ban non-`'static` data from them.
+
+### This Proposal Will:
+
+ - Allow types to manually opt-in to not being leaked.
+ - Allow leaks of `'static` data in a large variety of ways, even if they do
+   not implement `Leak`.
+ - Prevent types which have opted-in to not being leaked from being used in
+   contexts where they might leak.
+ - Prevent panics in destructors from causing leaks or other memory unsafety.
+ - Allow `Rc` and `Arc` cycles, including when they contain non-`'static` data.
+
+# Detailed design
+
+Introduce a new default trait to `core`, `Leak`:
+
+```rust
+pub unsafe trait Leak { }
+unsafe impl Leak for .. { }
+```
+
+Change several `std` APIs to adjust for the guarantees now provided to types
+which do not implement `Leak`:
+
+ - `impl<'a, T> !Leak for ::std::thread::JoinGuard<'a, T> { }`
+   - If we gain some form of specialization, implement `Leak` for `JoinGuard<'static, T>`
+ - implement `!Leak` for the guards returned by collection `drain` operations
+ - Add a `Leak` bound to the type parameter of:
+   - `rc::Rc`
+   - `sync::Arc`
+   - `sync::mpsc::Sender` (possibly not, see unresolved questions)
+   - Possibly other APIs. Please point any others out if you think of them.
+
+_Cause all panics in destructors to immediately abort the process._
```

Comment:
We have custom code in the standard library to handle a panic, so implementing this would require that code to be able to detect when the panic originated from a destructor or not. Answering the question of "did this panic originate from a destructor" may or may not be very expensive.

The simplest solution would be for all destructors to increment a thread-local flag when they start running and decrement it when they stop running. If this flag is greater than 0, then you know the panic originated while in some destructor somewhere. Modifying this flag from *all destructors*, however, seems like it would be *extremely* costly. I don't think that LLVM would have an easy time optimizing it all away, but this would need to be measured to confirm.

A non-native and 0-overhead solution might look like something along the lines of running the stack unwinder and if a frame of a destructor is encountered then an abort happens, otherwise the panic happens normally. I don't know how this would be implemented, nor if it could be implemented reliably in the face of inlining.